### PR TITLE
adds Writer.addAnnotation() and .setAnnotations(), and removes annotations param from write*() and stepIn() methods

### LINKED ISSUE #343: remove annotations parameter from Writer's write*Type*() and stepIn() methods
...in favor of ion-java's IonWriter `addAnnotation()`/`setAnnotations()` methods.  The current API is problematic when trying to set/add annotations while using `Writer.writeValues()` (see
[#319](https://github.com/amzn/ion-js/pull/319#issuecomment-526338662)).

The following code presents an example use case:
```javascript
let data = "{ x: 1 }";
let reader = ion.makeReader(data);
reader.next();

let writer = ion.makeTextWriter();
writer.setAnnotations(['a', 'b']);
writer.writeValue(reader);
writer.close();
// output:  a::b::{x: 1}
```

The `Writer` interface's approach to specifying annotations does not allow for this.

### PR DESCRIPTION
**Pull request by pbcornell**
*Description of changes:*
Moves `annotation` from `Writer.write*()` methods to `setAnnotations()` and `addAnnotation()` methods in a new `AbstractWriter` base class.

Also removes vestigial `isNull` parameter from a couple `writeContainer` methods (in TextWriter and PrettyTextWriter).

Resolves #343 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit/IonTextWriterTest.js [hunk 1]
```diff
@@ -367,41 +375,61 @@
     // PrettyPrint
     skippedPrettyTest('Writes composite pretty ion',
       writer => {
-        writer.stepIn(ion.IonTypes.STRUCT, ['a1']);
+        writer.setAnnotations(['a1']);
```
Comment (zslayton) [inline tests/unit/IonTextWriterTest.js]: Food for thought: maybe in a future release we could allow annotations to be specified either as an array or as [varargs-style strings](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters).
```typescript
writer.setAnnotations(['a1']);
writer.setAnnotations(['a1', 'a2', 'a3']);
//   vs.
writer.setAnnotations('a1');
writer.setAnnotations('a1', 'a2', 'a3');
```

File: src/AbstractWriter.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (pbcornell) [reply src/AbstractWriter.ts]: to be addressed by #303

File: src/util.ts [hunk 3]
```diff
@@ -53,22 +53,23 @@ export function _writeValue(reader: IonReader, writer: IonWriter, _depth = 0): v
             writer.writeFieldName(reader.fieldName());
         }
     }
+    writer.setAnnotations(reader.annotations());
```
[the remark this answers is not in the document]
Comment (pbcornell) [reply src/util.ts]: opened #371 for you

File: tests/unit/IonTextWriterTest.js (hunk 1 above)
Comment (pbcornell) [reply tests/unit/IonTextWriterTest.js]: opened #372 for you

**Pull request merged**